### PR TITLE
ENH: Add newly proposed colormaps

### PR DESCRIPTION
**Pull request by stefanv**
This PR adds four new color maps to matplotlib, including the newly proposed default `viridis`.

The colormaps are currently specified by 256 data values.

### CODE REVIEW

**Comment by tacaswell** [comment]
hurray!

**Comment by efiring** [comment]
Thank you!  
At some point it would be nice to condense the increasingly monstrous _cm.py; it has a lot of useless precision.  In many cases the LinearSegmentedColormap machinery is not needed.  In the end everything is being done with a lookup table, usually with 256 entries.  For something like these new colormaps, the only benefit of the LinearSegmentedColormap route is that it allows the generation of a lookup table with a larger number of entries.

File: boilerplate.py [hunk 1]
```diff
@@ -288,10 +288,14 @@ def format_value(value):
         'hot',
         'hsv',
         'jet',
+        'neoncity',
         'pink',
         'prism',
         'spring',
         'summer',
+        'sunset',
```
Comment (efiring) [inline boilerplate.py]: The advantage of "sunrise" over "sunset" is that the map goes from dark to light; but maybe one of the other naming schemes will override all this anyway.  (Also, "sunrise" has a more upbeat, optimistic feel...)

**Comment by tacaswell** [comment]
I am in favor of making one of the 'rocket' :rocket: per @mwaskom

File: lib/matplotlib/pyplot.py [hunk 2]
```diff
@@ -1971,12 +1971,16 @@ def colormaps():
                   by changing the hue component in the HSV color space
       jet         a spectral map with dark endpoints, blue-cyan-yellow-red;
                   based on a fluid-jet simulation by NCSA [#]_
+      neoncity    perceptually uniform shares of blue-red-yellow
```
Comment (WeatherGod) [inline lib/matplotlib/pyplot.py]: "shares" --> "shades"

**Comment by tacaswell** [comment]
@efiring Can you be in charge of the naming issues on our side?  Hopefully this thread stays calmer than the skimage thread.

**Comment by efiring** [comment]
OK; but just as there was no agreement among those polled about the virtues of A, B, and C, I expect virtually no agreement about how to name A, B, and C.  We'll work it out, though.

**Comment by WeatherGod** [comment]
I am fine with the names as-is.

**Comment by stefanv** [comment]
The only vague agreement I could sense in the other thread was for 'rocket'.  Shall I replace viola with rocket?

**Comment by efiring** [comment]
@Stefanv, I'm going to make a strong recommendation, with my rationale.  If you or Nathaniel are strongly opposed, then I will defer to you; otherwise, I hope my recommendation will prove in the long run to be a satisfactory way to get the paint on the bikeshed and move on.  I expect my recommendation will disappoint and/or annoy most of the people who have weighed in on this topic, but so be it.

General criteria for *most* non-default colormaps:
1) Names should have some visual connection to the color sequence in the map, and the closer, the better.
2) Names should be short and simple.
3) In case of subtle variations on a theme, ideally the naming should clearly indicate the grouping.

Starting with (1), the viable candidates are related to fire (I will put "rocket" in this category), magma, and sunrise/sunset.  All of these can have some near-black, all can have some very light yellows, and some reds and orange tones.  The problem with the first two of these is that I don't think they have magenta shades, which are prominent in A, B, and C.  To my eye, the sunrise/sunset names are by far the closest to the actual shades in the maps.  Sunrise has the additional advantage of going in the right direction, dark to light.

(2) eliminates suggestions like "softsunrise".

(3) with (2) favors the use of numbers to distinguish subtle variations.

My recommendation:
- B is "sunrise"
- A is "sunrise1"
- C is "sunrise2"

I gave B the priority of having no number because I think it is (by a whisker) the best default of the group: C is too blue, A lacks contrast with a white background at the high end.

**Comment by njsmith** [comment]
My personal favorite is still this: https://github.com/scikit-image/scikit-image/pull/1599#issuecomment-121747715
but I'll leave it at that :-)

**Comment by efiring** [comment]
@stefanv What do you think?  Please either pick my suggestion, or any of @njsmith's variations, or whatever you  negotiate with Nathaniel, and make the change to the PR.  I will merge it with pleasure.  No problem, no regrets, no more discussion needed.

**Comment by njsmith** [comment]
@stefanv: just on the off chance you've been waiting to catch me at the
office or something: I'm out of town all week, but not terribly busy so if
you do want my input on anything feel free to call or whatever. Or not :-).

**Comment by stefanv** [comment]
@efiring Is this the most optimal way of representing the colormap values?  I put a very large table in there, but I can also do it differently (I just followed what was in the file already).

I'll chat to Nathaniel and figure out what to do with the naming

**Comment by efiring** [comment]
Short answer: please use your list of rgb values to initialize a ListedColormap; and ideally, reduce the number of digits printed out in the list.

@stefanv,  in fact I don't like the way that _cm.py has ballooned even before this PR.  I think the LinearSegmentedColormap initialization via dictionaries for r, g, and b is highly inefficient and awkward for all of the maps that are being specified with a large number of points.  The specification can be cut down to 33% simply by feeding the rgb values into the `from_list()` method that will generate the same dictionary that you have now.  To make it even more efficient, use the list of rgb tuples to initialize a ListedColormap instead of a LinearSegmentedColormap.  As far as I can see, there will be no difference in the end result, but the ListedColormap goes straight to the LUT instead of taking a roundabout route to the same LUT.  File size and parsing time can be cut down even more by reducing the number of significant figures in the rgb values to something reasonable.   LinearSegmentedColormap was designed to allow the creation of both smooth and discontinuous maps from dictionaries with *only a few* entries; it's purpose is to do the interpolation, and it does so in rgb space.  There is no point in using it when a dense table of rgb values has been created by other means.

**Comment by njsmith** [comment]
So looking at the code to make sure I understand what you mean -- it looks like `LinearSegmentedColormap` ends up just discretizing to 256 points (by default, but everyone uses the default), and if I'm counting right the PR is already generating 256 points, so we might as well just do `ListedColormap(array with shape (256, 3), name=name)`? And write down that array with, say, 4 digits of precision, because it's going to an 8-bit-per-channel monitor anyway? (I guess there do exist workflows that can handle a bit more color precision than that, but even 5 or 6 digits is enough for a full 16-bit workflow if I'm counting right.)

**Comment by efiring** [comment]
@njsmith, yes, exactly.

**Comment by efiring** [comment]
The travis failure is in the doc build, and is unrelated.

**Comment by stefanv** [comment]
Eric, I gave it a shot--let me know what you think.  Happy to rework as necessary.

File: lib/matplotlib/pyplot.py [hunk 3]
```diff
@@ -1971,12 +1971,16 @@ def colormaps():
                   by changing the hue component in the HSV color space
       jet         a spectral map with dark endpoints, blue-cyan-yellow-red;
                   based on a fluid-jet simulation by NCSA [#]_
+      neoncity    perceptually uniform shades of blue-red-yellow
```
Comment (efiring) [inline lib/matplotlib/pyplot.py]: I think that updating the names in this docstring is the last thing remaining.

**Comment by stefanv** [comment]
Good catch!

**Comment by efiring** [comment]
PEP8 is being pesky as usual.  It is unhappy about the space after each '[' and before each ']'.

**Comment by stefanv** [comment]
Updated to fix spacing

**Comment by efiring** [comment]
Github's presentation of what has passed and what has failed has been changed recently, and the new version is self-contradictory, with two inconsistent statements, both of which are wrong.  In any case, looking at Travis, there are two failures.  One failure is the typical false alarm (random glitch associated with LaTeX), but pep8 still has one complaint: in _cm_listed.py, it wants all imports to be at the top of the file.  Sorry I missed that the last time.

**Comment by stefanv** [comment]
*grumble* I passed this through flake8 beforehand--which checker do you use?

**Comment by stefanv** [comment]
OK, let's try again!

**Comment by efiring** [comment]
pep8 testing is done with the pep8 module, invoked in lib/matplotlib/tests/test_coding_standards.py.  That, in turn, can be invoked from the tests.py script in the base directory, or via 
```
python setup.py test --pep8-only
```

**Comment by efiring** [comment]
I don't know why github says a check failed; I looked at the travis page, and everything was green.  In any case, the 4 new maps are now in place.  Thanks very much for all the work you and @njsmith have done leading up to this.

**Comment by stefanv** [comment]
Hah, interesting--this update was made to the PEP8 checker since my last install (a few months ago).  Thanks for the explanation and the reviews, Eric!

File: lib/matplotlib/pyplot.py [hunk 4]
```diff
@@ -3787,4 +3791,60 @@ def spectral():
     if im is not None:
         im.set_cmap(cm.spectral)
 
+
+# This function was autogenerated by boilerplate.py.  Do not edit as
+# changes will be lost
+def magma():
```
Comment (endolith) [inline lib/matplotlib/pyplot.py]: I thought these functions were not supposed to be created any longer?  Only the ones included for Matlab compatibility should exist, and all others should be used as `set_cmap('magma')`?  See https://github.com/matplotlib/matplotlib/pull/889#discussion_r867101

File: lib/matplotlib/cm.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
 import matplotlib.cbook as cbook
 from matplotlib._cm import datad
 from matplotlib._cm import cubehelix
+from matplotlib._cm_listed import cmaps as cmaps_listed
```
Comment (endolith) [inline lib/matplotlib/cm.py]: So cmaps created with this method will not have reversed variants like `viridis_r`?

**Comment by endolith** [comment]
@efiring 

> in fact I don't like the way that _cm.py has ballooned even before this PR. I think the LinearSegmentedColormap initialization via dictionaries for r, g, and b is highly inefficient and awkward for all of the maps that are being specified with a large number of points.

Should all the ColorBrewer maps be converted to ListedColormap based on the original values from the spreadsheet, then?  (Defined using 12 points per colormap at most.)

> LinearSegmentedColormap was designed to allow the creation of both smooth and discontinuous maps from dictionaries with only a few entries; it's purpose is to do the interpolation, and it does so in rgb space.

Should it interpolate in a perceptual space instead?

File: lib/matplotlib/pyplot.py (hunk 4 above)
Comment (jenshnielsen) [inline lib/matplotlib/pyplot.py]: There are auto generated so all that is done here is to regenerate pyplot with boilerplate.py 

If you want to change that I suggest raising  a new issue.

**Comment by OrkoHunter** [comment]
Is it true that [the plot](https://pbs.twimg.com/media/Ca8jlVIWcAUmeP8.png) used in the new [paper of the gravitational field](https://physics.aps.org/featured-article-pdf/10.1103/PhysRevLett.116.061102) used the color map added in this PR ? It looks so, does it?

**Comment by pelson** [comment]
> Is it true that the plot used in the new paper of the gravitational field used the color map added in this PR ? It looks so, does it?

Yes. The tutorial demonstrates it here: https://losc.ligo.org/s/events/GW150914/GW150914_tutorial.html.

**Comment by duncanmmacleod** [comment]
Yes, the LSC has been pretty good early adopters of the new defaults for
matplotlib 2.0, including this now famous figure.

Independent of colour maps, a significant fraction of the figures in this
paper and in the so-called 'companion' papers released on the arXiv today
use matplotlib for figure generation, with a scattering of matlab and
mathemetica in there as well. So, thanks go to the matplotlib community for
a great resource with which to present our work.


Duncan

On 12 February 2016 at 01:39, Phil Elson <notifications@github.com> wrote:

> Is it true that the plot used in the new paper of the gravitational field
> used the color map added in this PR ? It looks so, does it?
>
> Yes. The tutorial demonstrates it here:
> https://losc.ligo.org/s/events/GW150914/GW150914_tutorial.html.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4707#issuecomment-183216959>
> .
>



-- 
Duncan Macleod
duncan.macleod@ligo.org

**Comment by ghattab** [comment]
A small shout here! Any suggestions on how we should cite `viridis`?

**Pull request merged**